## Note: Frequency encoding was performed as part of the EDA in a previous notebook

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
df = pd.read_csv('../data/in_process/Games_FreqEncoded.csv', index_col = 'game_id')

In [3]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split


In [4]:
df.tail(2)

min_players  max_players  avg_time  year  avg_rating  geek_rating  \
game_id                                                                      
286096             1            5     120.0  2019     7.49800      7.14207   
287954             2            4      45.0  2019     7.84646      7.01513   

         age  weight  Area Majority / Influence  Auction/Bidding  ...  \
game_id                                                           ...   
286096    12  2.8564                        0.0              0.0  ...   
287954     8  2.0132                        0.0              0.0  ...   

         Video Game Theme  unknowncat_  Environmental  Pike and Shot  Medical  \
game_id                                                                         
286096                0.0          0.0            0.0            0.0      0.0   
287954                0.0          0.0            0.0            0.0      0.0   

         Vietnam War  Math  Mature / Adult  Korean War  \
game_id                                                  
286096           0.0   0.0             0.0         0.0   
287954           0.0   0.0             0.0         0.0   

         Expansion for Base-game  
game_id                           
286096                       0.0  
287954                       0.0  

[2 rows x 276 columns]

In [5]:
df.shape

(3836, 276)

In [6]:
df.iloc[:, :8].describe()

#Columns 8 through 276 are Frequency encoded categorical variables

min_players  max_players      avg_time         year   avg_rating  \
count  3836.000000  3836.000000   3836.000000  3836.000000  3836.000000   
mean      2.010688     5.436131     82.510167  2007.887904     6.834371   
std       0.681773     7.278325    273.188526    10.636066     0.698512   
min       1.000000     1.000000      1.000000  1960.000000     3.331770   
25%       2.000000     4.000000     30.000000  2004.000000     6.421880   
50%       2.000000     5.000000     60.000000  2011.000000     6.848635   
75%       2.000000     6.000000     90.000000  2015.000000     7.300432   
max       8.000000   100.000000  12000.000000  2020.000000     9.186900   

       geek_rating          age       weight  
count  3836.000000  3836.000000  3836.000000  
mean      6.222357    10.722888     2.217682  
std       0.539138     2.427055     0.782007  
min       4.272580     2.000000     1.000000  
25%       5.844862     8.000000     1.590000  
50%       6.109750    10.000000     2.150900  
75%       6.523722    12.000000     2.730650  
max       8.576860    18.000000     4.723300

In [7]:
#Dependant variables are average rating and "geek rating"
#These are two different rating scales. They should be rather similar, so we'll use Average Rating for now
#The rest are possible inputs

X = df.drop(['avg_rating', 'geek_rating'], axis = 1)
y = df['avg_rating']
y2 = df['geek_rating']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .25, random_state = 42)


In [8]:
X_train.head()

min_players  max_players  avg_time  year  age  weight  \
game_id                                                          
404                3            5      20.0  1994    7  1.1000   
53093              2            2      60.0  2014   14  2.7537   
3510               2            8      45.0  1997   12  1.0833   
1705               2           10      30.0  1982    8  1.3211   
24480              2            4     120.0  2006   12  2.8244   

         Area Majority / Influence  Auction/Bidding  Dice Rolling  \
game_id                                                             
404                            0.0              0.0      0.000000   
53093                          0.0              0.0      0.073673   
3510                           0.0              0.0      0.000000   
1705                           0.0              0.0      0.000000   
24480                          0.0              0.0      0.000000   

         Hand Management  ...  Video Game Theme  unknowncat_  Environmental  \
game_id                   ...                                                 
404             0.000000  ...               0.0          0.0            0.0   
53093           0.080517  ...               0.0          0.0            0.0   
3510            0.000000  ...               0.0          0.0            0.0   
1705            0.000000  ...               0.0          0.0            0.0   
24480           0.000000  ...               0.0          0.0            0.0   

         Pike and Shot  Medical  Vietnam War  Math  Mature / Adult  \
game_id                                                              
404                0.0      0.0          0.0   0.0             0.0   
53093              0.0      0.0          0.0   0.0             0.0   
3510               0.0      0.0          0.0   0.0             0.0   
1705               0.0      0.0          0.0   0.0             0.0   
24480              0.0      0.0          0.0   0.0             0.0   

         Korean War  Expansion for Base-game  
game_id                                       
404             0.0                      0.0  
53093           0.0                      0.0  
3510            0.0                      0.0  
1705            0.0                      0.0  
24480           0.0                      0.0  

[5 rows x 274 columns]

In [9]:
X_train.iloc[:,6:] #columns with categorical variables


Area Majority / Influence  Auction/Bidding  Dice Rolling  \
game_id                                                             
404                       0.000000              0.0      0.000000   
53093                     0.000000              0.0      0.073673   
3510                      0.000000              0.0      0.000000   
1705                      0.000000              0.0      0.000000   
24480                     0.000000              0.0      0.000000   
...                            ...              ...           ...   
19995                     0.000000              0.0      0.000000   
27389                     0.000000              0.0      0.073673   
8229                      0.039385              0.0      0.000000   
232043                    0.000000              0.0      0.000000   
199383                    0.039385              0.0      0.000000   

         Hand Management  Simultaneous Action Selection  Trick-taking  \
game_id                                                                 
404             0.000000                       0.017737           0.0   
53093           0.080517                       0.000000           0.0   
3510            0.000000                       0.000000           0.0   
1705            0.000000                       0.017737           0.0   
24480           0.000000                       0.000000           0.0   
...                  ...                            ...           ...   
19995           0.000000                       0.000000           0.0   
27389           0.000000                       0.000000           0.0   
8229            0.000000                       0.000000           0.0   
232043          0.000000                       0.000000           0.0   
199383          0.000000                       0.000000           0.0   

         Set Collection  Tile Placement  Investment  Market  ...  \
game_id                                                      ...   
404                 0.0        0.000000         0.0     0.0  ...   
53093               0.0        0.000000         0.0     0.0  ...   
3510                0.0        0.000000         0.0     0.0  ...   
1705                0.0        0.000000         0.0     0.0  ...   
24480               0.0        0.000000         0.0     0.0  ...   
...                 ...             ...         ...     ...  ...   
19995               0.0        0.027863         0.0     0.0  ...   
27389               0.0        0.000000         0.0     0.0  ...   
8229                0.0        0.000000         0.0     0.0  ...   
232043              0.0        0.027863         0.0     0.0  ...   
199383              0.0        0.000000         0.0     0.0  ...   

         Video Game Theme  unknowncat_  Environmental  Pike and Shot  Medical  \
game_id                                                                         
404                   0.0          0.0            0.0            0.0      0.0   
53093                 0.0          0.0            0.0            0.0      0.0   
3510                  0.0          0.0            0.0            0.0      0.0   
1705                  0.0          0.0            0.0            0.0      0.0   
24480                 0.0          0.0            0.0            0.0      0.0   
...                   ...          ...            ...            ...      ...   
19995                 0.0          0.0            0.0            0.0      0.0   
27389                 0.0          0.0            0.0            0.0      0.0   
8229                  0.0          0.0            0.0            0.0      0.0   
232043                0.0          0.0            0.0            0.0      0.0   
199383                0.0          0.0            0.0            0.0      0.0   

         Vietnam War  Math  Mature / Adult  Korean War  \
game_id                                                  
404              0.0   0.0             0.0         0.0   
53093            0.0   0.0             0.0      

In [10]:
#Scale just the numerical part of the training set. Leaving the categoricals alone
XScaled = X_train.iloc[:,:6]
col_names = XScaled.columns
XCats = X_train.iloc[:,6:] #columns with categorical variables
scaler = StandardScaler()
XScaled = scaler.fit_transform(XScaled)
X_train = pd.DataFrame(XScaled, columns = col_names, index = X_train.index)
X_train = X_train.join(XCats, how = 'left') #merge the scaled columns with the encoded columns
X_train.tail()

min_players  max_players  avg_time      year       age    weight  \
game_id                                                                     
19995       0.002060    -0.058063  0.044059 -0.191676  0.530695  0.871417   
27389       0.002060     0.365198 -0.237885 -0.191676  0.530695 -1.123985   
8229        0.002060    -0.199150 -0.237885 -0.474665 -1.115016 -0.325080   
232043      0.002060    -0.199150 -0.261380  0.845952 -1.115016 -0.149143   
199383      1.484016    -0.058063 -0.026427  0.845952 -0.292161  0.828653   

         Area Majority / Influence  Auction/Bidding  Dice Rolling  \
game_id                                                             
19995                     0.000000              0.0      0.000000   
27389                     0.000000              0.0      0.073673   
8229                      0.039385              0.0      0.000000   
232043                    0.000000              0.0      0.000000   
199383                    0.039385              0.0      0.000000   

         Hand Management  ...  Video Game Theme  unknowncat_  Environmental  \
game_id                   ...                                                 
19995                0.0  ...               0.0          0.0            0.0   
27389                0.0  ...               0.0          0.0            0.0   
8229                 0.0  ...               0.0          0.0            0.0   
232043               0.0  ...               0.0          0.0            0.0   
199383               0.0  ...               0.0          0.0            0.0   

         Pike and Shot  Medical  Vietnam War  Math  Mature / Adult  \
game_id                                                              
19995              0.0      0.0          0.0   0.0             0.0   
27389              0.0      0.0          0.0   0.0             0.0   
8229               0.0      0.0          0.0   0.0             0.0   
232043             0.0      0.0          0.0   0.0             0.0   
199383             0.0      0.0          0.0   0.0             0.0   

         Korean War  Expansion for Base-game  
game_id                                       
19995           0.0                      0.0  
27389           0.0                      0.0  
8229            0.0                      0.0  
232043          0.0                      0.0  
199383          0.0                      0.0  

[5 rows x 274 columns]

In [11]:
X_test.iloc[:,:6]

min_players  max_players  avg_time  year  age  weight
game_id                                                       
19370              3            8      60.0  2005   12  1.2793
28037              2            6      30.0  2007   10  1.4925
6778               2            4      90.0  2003   12  2.5400
21892              2            5      75.0  2006   12  2.6708
169786             1            5     115.0  2016   14  3.3881
...              ...          ...       ...   ...  ...     ...
21985              2            5      30.0  2006    6  1.1944
1855               3           10      60.0  1991    8  1.0952
3655               2            4      60.0  2002    7  2.3277
126239             2            6      50.0  2013   10  2.4103
210232             1            4     360.0  2017   10  3.5319

[959 rows x 6 columns]

In [12]:
#Now we need to scale the Test Set using the same scaler we used on the training set
#Scale just the first 6 columns of test set. The rest are categorical
XScaled = X_test.iloc[:,:6]
col_names = XScaled.columns
XCats = X_test.iloc[:,6:] #columns with categorical variables
XScaled = scaler.transform(XScaled)
X_test = pd.DataFrame(XScaled, columns = col_names, index = X_test.index)
X_test = X_test.join(XCats, how = 'left') #merge the scaled columns with the hot encoded columns
X_test.tail()


min_players  max_players  avg_time      year       age    weight  \
game_id                                                                     
21985       0.002060    -0.058063 -0.237885 -0.191676 -1.937872 -1.318928   
1855        1.484016     0.647372 -0.096913 -1.606623 -1.115016 -1.446321   
3655        0.002060    -0.199150 -0.096913 -0.568995 -1.526444  0.136465   
126239      0.002060     0.083024 -0.143904  0.468633 -0.292161  0.242540   
210232     -1.479895    -0.199150  1.312806  0.845952 -0.292161  1.682908   

         Area Majority / Influence  Auction/Bidding  Dice Rolling  \
game_id                                                             
21985                          0.0              0.0      0.000000   
1855                           0.0              0.0      0.000000   
3655                           0.0              0.0      0.000000   
126239                         0.0              0.0      0.000000   
210232                         0.0              0.0      0.073673   

         Hand Management  ...  Video Game Theme  unknowncat_  Environmental  \
game_id                   ...                                                 
21985           0.000000  ...               0.0          0.0            0.0   
1855            0.000000  ...               0.0          0.0            0.0   
3655            0.080517  ...               0.0          0.0            0.0   
126239          0.000000  ...               0.0          0.0            0.0   
210232          0.000000  ...               0.0          0.0            0.0   

         Pike and Shot  Medical  Vietnam War  Math  Mature / Adult  \
game_id                                                              
21985              0.0      0.0          0.0   0.0             0.0   
1855               0.0      0.0          0.0   0.0             0.0   
3655               0.0      0.0          0.0   0.0             0.0   
126239             0.0      0.0          0.0   0.0             0.0   
210232             0.0      0.0          0.0   0.0             0.0   

         Korean War  Expansion for Base-game  
game_id                                       
21985           0.0                      0.0  
1855            0.0                      0.0  
3655            0.0                      0.0  
126239          0.0                      0.0  
210232          0.0                      0.0  

[5 rows x 274 columns]

In [13]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.linear_model import LinearRegression, Lasso, LassoCV, RidgeCV
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestRegressor

### Trying models on scaled and frequency encoded data

In [14]:
#Trying a strait linear regression on all columns as a reference
lm = LinearRegression()
lm.fit(X_train, y_train)
y_pred = lm.predict(X_test)

print('MSE = ', mean_squared_error(y_test, y_pred))
print('R^2 = ', r2_score(y_test, y_pred))

MSE =  0.22861230798774104
R^2 =  0.5188154888389083


In [15]:
#Lasso Regression performed with .01 as reference
alpha = .01
lassom = Lasso(alpha = alpha)

lassom.fit(X_train, y_train)
y_pred_lasso = lassom.predict(X_test)

print('MSE = ', mean_squared_error(y_test, y_pred_lasso))
print('R^2 = ', r2_score(y_test, y_pred_lasso))

MSE =  0.2414009893396125
R^2 =  0.49189779819110113


In [16]:
print(lassom.sparse_coef_)
bestCoefs = lassom.coef_
features = X_train.columns
pd.Series(bestCoefs, index=features).sort_values(ascending=False)

  (0, 0)	-0.017826213908531296
  (0, 2)	0.007657361623842103
  (0, 3)	0.25906356811076753
  (0, 4)	0.00810477033363821
  (0, 5)	0.35949456935047


weight                          0.359495
year                            0.259064
age                             0.008105
avg_time                        0.007657
Singing                        -0.000000
                                  ...   
Resource to Move                0.000000
Mancala                        -0.000000
Real-Time                       0.000000
Victory Points as a Resource    0.000000
min_players                    -0.017826
Length: 274, dtype: float64

In [17]:
#Finding the best coefficient for Alpha
lassoc = LassoCV()
lassoc.fit(X_train,y_train)
print(lassoc.alpha_)
pd.Series(lassoc.coef_, index=features).sort_values(ascending=False)


0.0053511991683924225


weight                          0.360389
year                            0.263049
avg_time                        0.011888
age                             0.010574
Singing                        -0.000000
                                  ...   
Resource to Move                0.000000
Mancala                        -0.000000
Real-Time                       0.000000
Victory Points as a Resource    0.000000
min_players                    -0.021389
Length: 274, dtype: float64

In [18]:
y_pred_lassoc = lassoc.predict(X_test)

In [19]:
print('MSE lassoCV = ', mean_squared_error(y_test, y_pred_lassoc))
print('R^2 lassoCV = ', r2_score(y_test, y_pred_lassoc))

MSE lassoCV =  0.24163291817221233
R^2 lassoCV =  0.49140963303970997


In [20]:
#Using grid search as an alternate to find best Alpha for Lasso Regression
param_grid = {'alpha':np.arange(.004, .051, .001)}
lassom2 = Lasso()
lasso_cv = GridSearchCV(lassom2,param_grid,cv=5, n_jobs =-1)
lasso_cv.fit(X_train,y_train)

print("Best Score:" + str(lasso_cv.best_score_))
print("Best Parameters: " + str(lasso_cv.best_params_))

Best Score:0.47820811348839537
Best Parameters: {'alpha': 0.006}


In [21]:
#Trying a random forest regression with random regressors search
rf = RandomForestRegressor(random_state = 42)
rnd_param_grid = {'n_estimators':list(np.arange(4, 270))}
rfcv = RandomizedSearchCV(rf, rnd_param_grid, n_jobs = -1)
search = rfcv.fit(X_train, y_train)
search.best_params_

{'n_estimators': 269}

In [22]:
print(search.best_estimator_)
print(search.best_score_)
print(search.best_index_)

RandomForestRegressor(n_estimators=269, random_state=42)
0.5793818838538545
6


In [23]:
#Narrowing down with grid search
srch_param_grid = {'n_estimators':(np.arange(260, 270))}
gscv = GridSearchCV(rf, srch_param_grid, n_jobs = -1)
search2 = gscv.fit(X_train, y_train)
search2.best_params_

{'n_estimators': 265}

In [24]:
print(search2.best_estimator_, search2.best_score_)

RandomForestRegressor(n_estimators=265, random_state=42) 0.5793897175527123


In [25]:
y_pred_rf = search2.predict(X_test)
print('MSE = ', mean_squared_error(y_test, y_pred_rf))
print('R^2 = ', r2_score(y_test, y_pred_rf))

MSE =  0.19444961436306663
R^2 =  0.5907213244276754


In [26]:
y_pred_lassocv = lasso_cv.predict(X_test)

In [27]:
ridgemodel = RidgeCV(alphas = [.0001, .001, .01, 1])
ridgemodel.fit(X_train, y_train)
print(ridgemodel.alpha_)
print(ridgemodel.best_score_)
pd.Series(ridgemodel.coef_, index=features).sort_values(ascending=False)

0.0001
-0.2302423129765688


Race                    88.385153
Communication Limits    77.710713
Line Drawing            69.808202
Traitor Game            64.607265
Sports                  59.849325
                          ...    
Video Game Theme       -42.040301
Memorycat_             -46.256849
Zombies                -49.085767
Rock-Paper-Scissors    -65.043748
Trivia                 -76.510057
Length: 274, dtype: float64

In [28]:
y_pred_ridge = ridgemodel.predict(X_test)


In [29]:
print('MSE simple linear = ', mean_squared_error(y_test, y_pred))
print('R^2 simple linear = ', r2_score(y_test, y_pred))
print('MSE RF = ', mean_squared_error(y_test, y_pred_rf))
print('R^2 RF = ', r2_score(y_test, y_pred_rf))
print('MSE LassoCV = ', mean_squared_error(y_test, y_pred_lassocv))
print('R^2 LassoCV = ', r2_score(y_test, y_pred_lassocv))
print('MSE lasso grid search = ', mean_squared_error(y_test, y_pred_lassoc))
print('R^2 lasso grid search = ', r2_score(y_test, y_pred_lassoc))
print('MSE RidgeCV = ', mean_squared_error(y_test, y_pred_ridge))
print('R^2 RidgeCV = ', r2_score(y_test, y_pred_ridge))

MSE simple linear =  0.22861230798774104
R^2 simple linear =  0.5188154888389083
MSE RF =  0.19444961436306663
R^2 RF =  0.5907213244276754
MSE LassoCV =  0.2415865644630485
R^2 LassoCV =  0.4915071985954791
MSE lasso grid search =  0.24163291817221233
R^2 lasso grid search =  0.49140963303970997
MSE RidgeCV =  0.22172086184745826
R^2 RidgeCV =  0.5333206446259833


The RandomForest regression gives the best scores, but it uses 265 of the features.
Ridge regression also gives a better score, but it's also using all the features. I question its predictive value.
The conclusion of the Lasso regression is that the Categorical features are meaningless.
I tend to like this. It shows that there's no magic bullet in game design. A game has to be good, independent of its mechanics.

QUESTION: How to interpret ridge regression?